### PR TITLE
support resource-limit

### LINKED ISSUE #51: [feature] Support resource limits
/kind feature

We need to support resource limits using magic commands or config files. I prefer to use config files first then support overwriting it using magic commands.

### PR DESCRIPTION
**Pull request by yeya24**
**What this PR does / why we need it**:

Add support for resource-limit. Currently only support cpu and memory.

There are two ways to set resource of one jobs.
- From config file, which is default.
- From magic commands, which can override the default config.

For the first way, we can set config file like this.
```yaml
kubeconfig: {path to your kubeconfig}
namespace: test
s2i:
  provider: s2i
worker:
  cpu: {default-worker-cpu}
  memory: {default-worker-memory}
ps:
  cpu: {default-ps-cpu}
  memory: {default-ps-memory}
master:
  cpu: {default-master-cpu}
  memory: {default-master-memory}      
```
For the second way, add two magic commands %cpu, %memory.
Here is an example. By this config we can start a ps of 0.5 core and 0.5G, and a worker of 1c and 1G.
``` 
%framework=tensorflow
%ps=1;%cpu=500m;%memory=500Mi
%worker=1;%cpu=1000m;%memory=1Gi
```

I dont know if it is a right way for this,  I' d appreciate if you give me some advice. :smile: 

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #51 

**Special notes for your reviewer**:

/cc @gaocegege 

**Release note**:

```release-note
Add support for resource-limit
```

### CODE REVIEW

File: cmd/kubeflow-kernel/command/run.go [hunk 1]
```diff
@@ -112,3 +115,22 @@ func createBackend(s2iConfig map[string]string, kubeconfig *restclientset.Config
...
+		res.PsCpu = psResource[config.CPU]
+		res.PsMemory = psResource[config.Memory]
+	}
+	if masterResource := viper.GetStringMapString(config.Master); masterResource != nil {
+		res.MasterCpu = masterResource[config.CPU]
+		res.MasterMemory = masterResource[config.Memory]
+	}
+
+	return *res
+}
```
Comment (gaocegege) [inline cmd/kubeflow-kernel/command/run.go]: Please add a newline here.

File: pkg/backend/kubeflow/generator/cm_test.go [hunk 2]
```diff
@@ -35,19 +39,41 @@ func TestCMNewTFJob(t *testing.T) {
...
 
 	param := &types.Parameter{
 		PSCount:     expectedPSCount,
 		WorkerCount: expectedWorkerCount,
 		Image:       expectedCM,
 		CleanPolicy: types.CleanPodPolicyAll,
+		Resource: resource.JobResource{
+			WorkerCpu:    "1000m",
+			WorkerMemory: "1Gi",
+			PsCpu:        "100m",
```
Comment (gaocegege) [inline pkg/backend/kubeflow/generator/cm_test.go]: PSCPU looks promising here.

File: pkg/backend/kubeflow/generator/cm_test.go [hunk 3]
```diff
@@ -35,19 +39,41 @@ func TestCMNewTFJob(t *testing.T) {
...
+			PsMemory:     "100Mi",
```
Comment (gaocegege) [inline pkg/backend/kubeflow/generator/cm_test.go]: PSMemory

File: pkg/config/config.go [hunk 4]
```diff
@@ -27,4 +27,11 @@ const (
 	S2IRegistry    = "registry"
 	S2IUsername    = "username"
 	S2IPassword    = "password"
+
+	Worker = "worker"
+	Ps     = "ps"
```
Comment (gaocegege) [inline pkg/config/config.go]: s/Ps/PS

File: pkg/interpreter/simple/simple.go [hunk 5]
```diff
@@ -19,32 +19,43 @@ import (
...
+)
+
 // Interpreter is the type for the simple interpreter.
 type Interpreter struct {
 	FrameworkPrefix   string
 	WorkerPrefix      string
 	PSPrefix          string
 	MasterPrefix      string
 	CleanPolicyPrefix string
+	CpuPrefix         string
```
Comment (gaocegege) [inline pkg/interpreter/simple/simple.go]: s/Cpu/CPU

File: pkg/resource/resource_test.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+		actual, err := testCase.Resource.MasterLimits()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if !reflect.DeepEqual(actual, testCase.Expected) {
+			t.Errorf("Expected %v, got %v", testCase.Expected, actual)
+		}
+	}
+}
```
Comment (gaocegege) [inline pkg/resource/resource_test.go]: Please add a newline here.

**Comment by yeya24** [comment]
@gaocegege  PTAL.

**Pull request merged**